File: pkg/automaton/automaton.go
package automaton

import (
	"errors"
	"fmt"
)

type Transition struct {
    FromState string
    Symbol    rune
    ToState   string
}

type Automaton struct {
    States       []string
    Alphabet     []rune

    Transitions  []Transition
    StartState   string
    AcceptStates []string
}

func containsRune(arr []rune, str rune) bool {
    for _, a := range arr {
        if a == str {
            return true
        }
    }
    return false
}

func contains(arr []string, str string) bool {
    for _, a := range arr {
        if a == str {
            return true
        }
    }
    return false
}

func isStatesValid(states []string) error {
    for index, state := range states {
        if state == "" {
            return errors.New(fmt.Sprintf("State of index %d invalid, can not be empty", index))
        }
    }

    return nil
}

func isTransitionsValid(states []string, alphabet []rune, transitions []Transition) error {
    for index, transition := range transitions {
        if !contains(states, transition.FromState) {
            return errors.New(fmt.Sprintf("Transition of index %d is invalid! FromState '%s' is not in the list of states", index, transition.FromState))
        } else if !containsRune(alphabet, transition.Symbol) {
            return errors.New(fmt.Sprintf("Transition of index %d is invalid! Symbol '%c' is not in alphabet", index, transition.Symbol))
        } else if !contains(states, transition.ToState) {
            return errors.New(fmt.Sprintf("Transition of index %d is invalid! ToState '%s' is not in the list of states", index, transition.ToState))
        }
    }

    return nil
}

func isAcceptStatesValid(states []string, acceptStates []string) error {
    for _, state := range acceptStates {
        if !contains(states, state) {
            return errors.New(fmt.Sprintf("Accept state '%s' is not in the list of states", state))
        }
    }

    return nil
}

func NewAutomaton(states []string, alphabet []rune, transitions []Transition, startState string, acceptStates []string) (*Automaton, error) {

    errorStates := isStatesValid(states)
    if errorStates != nil {
        return nil, errorStates
    }

    errorTransitions := isTransitionsValid(states, alphabet, transitions)
    if errorTransitions != nil {
        return nil, errorTransitions
    }

    if !contains(states, startState) {
        return nil, errors.New(fmt.Sprintf("Start state '%s' is not in the list of states", startState))
    }

    errorAcceptStates := isAcceptStatesValid(states, acceptStates)
    if errorAcceptStates != nil {
        return nil, errorAcceptStates
    }

    return &Automaton{
        States:       states,
        Alphabet:     alphabet,
        Transitions:  transitions,
        StartState:   startState,
        AcceptStates: acceptStates,
    }, nil
}

func (aut Automaton) GetNextState(currentState string, symbol rune) (string, error) {
    for _, transition := range aut.Transitions {
        if transition.FromState == currentState && transition.Symbol == symbol {
            return transition.ToState, nil
        }
    }

    return "", errors.New(fmt.Sprintf("There is no transition for the state '%s' and symbol '%c'", currentState, symbol))
}

func (aut Automaton) CheckInput(input string) (bool, error) {
    currentState := aut.StartState

    for _, symbol := range input {
        if !containsRune(aut.Alphabet, symbol) {
            return false, errors.New(fmt.Sprintf("Symbol '%c' is not in the alphabet", symbol))
        }

        newState, err := aut.GetNextState(currentState, symbol)

        if err != nil {
            return false, err
        }

        currentState = newState
    }

    return contains(aut.AcceptStates, currentState), nil
}

File: src/automaton/automaton.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "automaton.h"

#define MAX_STATES 256
#define MAX_ALPHABET 36
#define MAX_TRANSITIONS MAX_STATES * MAX_ALPHABET

typedef struct STransition {
    char *from;
    char c;
    char *to;
} Transition;

struct SAutomaton {
    char *states[MAX_STATES];
    size_t statesCount;
    char alphabet[MAX_ALPHABET];
    size_t alphabetCount;
    Transition transitions[MAX_TRANSITIONS];
    size_t transitionsCount;
    char *startState;
    char *acceptStates[MAX_STATES];
    size_t acceptStatesCount;
};

int _stringInArray(char *array[], size_t size, char *value);
int _charInArray(char array[], size_t size, char value);
int _automatonStateExists(Automaton *automaton, char *state);
int _automatonSymbolExists(Automaton *automaton, char c);
char *_automatonGetNextState(Automaton *automaton, char *state, char c);
int _automatonCountTransitions(Automaton *automaton, char *state, char c);

/*****************************************************************************
*                              PUBLIC FUNCTIONS                              *
******************************************************************************/

Automaton *automatonCreate() {
    size_t len = sizeof(Automaton);

    Automaton *automaton = malloc(len);
    memset(automaton, 0, len);

    automaton->statesCount = 0;
    automaton->alphabetCount = 0;
    automaton->transitionsCount = 0;
    automaton->startState = NULL;
    automaton->acceptStatesCount = 0;

    return automaton;
}

void automatonDestroy(Automaton **automaton) {
    if (*automaton) {
        free(*automaton);
    }

    automaton = NULL;
}

void automatonAddState(Automaton *automaton, char *state) {
    if (automaton->statesCount >= MAX_STATES) {
        fprintf(stderr, "Error max size of states is %d\n", MAX_STATES);
        exit(EXIT_FAILURE);
    } else if (_automatonStateExists(automaton, state)) {
        fprintf(stderr, "Error state '%s' is already setted\n", state);
        exit(EXIT_FAILURE);
    }

    automaton->states[automaton->statesCount] = state;
    automaton->statesCount++;
}

void automatonAddToAlphabet(Automaton *automaton, char c) {
    if (automaton->alphabetCount >= MAX_ALPHABET) {
        fprintf(stderr, "Error max size of alphabet is %d\n", MAX_ALPHABET);
        exit(EXIT_FAILURE);
    } else if (_automatonSymbolExists(automaton, c)) {
        fprintf(stderr, "Error symbol '%c' is already in alphabet\n", c);
        exit(EXIT_FAILURE);
    }

    automaton->alphabet[automaton->alphabetCount] = c;
    automaton->alphabetCount++;
}

void automatonAddTransition(Automaton *automaton, char *from, char c, char *to) {
    if (automaton->transitionsCount >= MAX_TRANSITIONS) {
        fprintf(stderr, "Error max size of transitions is %d\n", MAX_TRANSITIONS);
        exit(EXIT_FAILURE);
    } else if (!_automatonStateExists(automaton, from)) {
        fprintf(stderr, "Error state '%s' does not exist\n", from);
        exit(EXIT_FAILURE);
    } else if (!_automatonSymbolExists(automaton, c)) {
        fprintf(stderr, "Error symbol '%c' does not exist in the alphabet\n", c);
        exit(EXIT_FAILURE);
    } else if (!_automatonStateExists(automaton, to)) {
        fprintf(stderr, "Error state '%s' does not exist\n", to);
        exit(EXIT_FAILURE);
    }

    Transition t;
    t.from = from;
    t.c = c;
    t.to = to;

    automaton->transitions[automaton->transitionsCount] = t;
    automaton->transitionsCount++;
}

void automatonValidateTransitions(Automaton *automaton) {
    for (size_t i = 0; i < automaton->statesCount; i++) {
        char *state = automaton->states[i];

        for (size_t j = 0; j < automaton->alphabetCount; j++) {
            char c = automaton->alphabet[j];
            int totalTransitions = _automatonCountTransitions(automaton, state, c);

            if (totalTransitions < 1) {
                fprintf(stderr, "Error transition from state '%s' with symbol '%c' does not exist\n", state, c);
                exit(EXIT_FAILURE);
            } else if (totalTransitions > 1) {
                fprintf(stderr, "Error a total of %d transitions were found from state '%s' with symbol '%c'\n", totalTransitions, state, c);
                exit(EXIT_FAILURE);
            }
        }

        int found = 0;
        for (size_t i = 0; i < automaton->transitionsCount; i++) {
            Transition t = automaton->transitions[i];

            if (strcmp(t.to, state) == 0) {
                found = 1;
            }
        }

        if (!found) {
            fprintf(stderr, "Error there is no transitions to state '%s'\n", state);
            exit(EXIT_FAILURE);
        }
    }
}

void automatonAddStartState(Automaton *automaton, char *state) {
    if (automaton->startState) {
        fprintf(stderr, "Error start state is already setted\n");
        exit(EXIT_FAILURE);
    } else if (!_automatonStateExists(automaton, state)) {
        fprintf(stderr, "Error state '%s' does not exist\n", state);
        exit(EXIT_FAILURE);
    }

    automaton->startState = state;
}

void automatonAddAcceptState(Automaton *automaton, char *state) {
    if (automaton->acceptStatesCount >= MAX_STATES) {
        fprintf(stderr, "Error max size of accept states is %d\n", MAX_STATES);
        exit(EXIT_FAILURE);
    }else if (!_stringInArray(automaton->states, automaton->statesCount, state)) {
        fprintf(stderr, "Error state '%s' does not exist\n", state);
        exit(EXIT_FAILURE);
    }

    automaton->acceptStates[automaton->acceptStatesCount] = state;
    automaton->acceptStatesCount++;
}

int automatonCheck(Automaton *automaton, char *input) {
    char *state = automaton->startState;

    for (size_t i = 0; i < strlen(input); i++) {
        char c = input[i];
        char *nextState;

        if (!_charInArray(automaton->alphabet, automaton->alphabetCount, c)) {
            return 0;
        } else if (!(nextState = _automatonGetNextState(automaton, state, c))) {
            return 0;
        }

        state = nextState;
    }

    return _stringInArray(automaton->acceptStates, automaton->acceptStatesCount, state);
}

/*****************************************************************************
*                              PRIVATE FUNCTIONS                             *
******************************************************************************/

int _stringInArray(char *array[], size_t size, char *value) {
    for (size_t i = 0; i < size; i++) {
        if (strcmp(array[i], value) == 0) {
            return 1;
        }
    }

    return 0;
}

int _charInArray(char array[], size_t size, char value) {
    for (size_t i = 0; i < size; i++) {
        if (array[i] == value) {
            return 1;
        }
    }

    return 0;
}

int _automatonStateExists(Automaton *automaton, char *state) {
    return _stringInArray(automaton->states, automaton->statesCount, state);
}

int _automatonSymbolExists(Automaton *automaton, char c) {
    return _charInArray(automaton->alphabet, automaton->alphabetCount, c);
}

char *_automatonGetNextState(Automaton *automaton, char *state, char c) {
    for (size_t i = 0; i < automaton->transitionsCount; i++) {
        Transition t = automaton->transitions[i];

        if (strcmp(t.from, state) == 0 && t.c == c) {
            return t.to;
        }
    }

    return NULL;
}

int _automatonCountTransitions(Automaton *automaton, char *state, char c) {
    int total = 0;

    for (size_t i = 0; i < automaton->transitionsCount; i++) {
        Transition t = automaton->transitions[i];

        if (strcmp(t.from, state) == 0 && t.c == c) {
            total++;
        }
    }

    return total;
}

